File: leetcode/src/3sum.cpp
/*
    Solution 1: using a hashmap
*/
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        unordered_map<int, int> map;
        vector<vector<int>> result;

        int N = nums.size();

        for(int i = 0; i < N ; i++) {
            map[nums[i]] = i;
        }

        for (int i = 0; i < N-2; i ++) {
            if (i != 0 && nums[i-1] == nums[i]) continue;
            int a = nums[i];

            for(int j = i+1; j < N; j ++) {
                if (j != i+1 && nums[j-1] == nums[j]) continue;
                int b = nums[j];
                int c = -a-b;


                if (map.find(c) != map.end() && map[c] > j) {
                    result.push_back({a,b,c});
                }
            }
        }

        return result;
    }
};

File: leetcode/src/3sum.py
"""
    Method 1: two pointers    
"""
class Solution:

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []

        nums.sort()
        for i in range(len(nums)-2):

            if (i != 0 and nums[i-1] == nums[i] and i < len(nums)-2):
                continue
            
            j = i+1
            k = len(nums)-1
            while (j < k):
                
                s = nums[i]+nums[j]+nums[k]
                if (s < 0):
                    j += 1
                    while nums[j-1] == nums[j] and j < k:
                        j += 1
                elif (s > 0):
                    k -= 1
                    while nums[k+1] == nums[k] and j < k:
                        k -= 1
                else:
                    result.append([nums[i],nums[j],nums[k]])
                    j += 1
                    while nums[j-1] == nums[j] and j < k:
                        j += 1
                    k -= 1
                    while nums[k+1] == nums[k] and j < k:
                        k -= 1
        return [list(i) for i in result]


